File: bench/04_collatz/bench.py
import time

def collatz_len(n):
    steps = 1
    while n != 1:
        if n % 2 == 0:
            n //= 2
        else:
            n = 3 * n + 1
        steps += 1
    return steps

t0 = time.time()
limit = 1_000_000
max_steps = 0
max_n = 0
for i in range(1, limit):
    s = collatz_len(i)
    if s > max_steps:
        max_steps = s
        max_n = i
ms = (time.time() - t0) * 1000
print(f"result: {max_n} steps={max_steps}  time: {ms:.0f}ms")

File: bench/04_collatz/bench.rs
use std::time::Instant;

fn collatz_len(mut n: u64) -> u64 {
    let mut steps = 1;
    while n != 1 {
        if n % 2 == 0 { n /= 2; } else { n = 3 * n + 1; }
        steps += 1;
    }
    steps
}

fn main() {
    let t0 = Instant::now();
    let (max_n, max_steps) = (1u64..1_000_000)
        .map(|i| (i, collatz_len(i)))
        .max_by_key(|&(_, s)| s)
        .unwrap();
    let ms = t0.elapsed().as_millis();
    println!("result: {} steps={}  time: {}ms", max_n, max_steps, ms);
}
